File: lib/src/features/monitoring/data/transport/ble_permissions.dart
import 'dart:io';

import 'package:permission_handler/permission_handler.dart';

/// Runtime permissions BLE scanning needs.
///
/// Android 12 (API 31) split Bluetooth into `BLUETOOTH_SCAN` / `CONNECT`;
/// below that, scanning counts as a location request. iOS handles this through
/// the Info.plist usage string and needs nothing here.
abstract final class BlePermissions {
  static Future<bool> ensureGranted() async {
    if (!Platform.isAndroid) return true;

    final statuses = await <Permission>[
      Permission.bluetoothScan,
      Permission.bluetoothConnect,
      Permission.locationWhenInUse,
    ].request();

    // locationWhenInUse is only required on Android 11 and below; treat it as
    // optional so newer devices are not blocked by a denial they never saw.
    final scan = statuses[Permission.bluetoothScan];
    final connect = statuses[Permission.bluetoothConnect];

    final scanOk = scan == null || scan.isGranted;
    final connectOk = connect == null || connect.isGranted;

    if (scanOk && connectOk) return true;

    final location = statuses[Permission.locationWhenInUse];
    return location != null && location.isGranted;
  }
}

File: react_native/src/features/monitoring/data/transport/blePermissions.ts
import { PermissionsAndroid, Platform } from 'react-native';

/**
 * Runtime permissions BLE scanning needs.
 *
 * Android 12 (API 31) split Bluetooth into `BLUETOOTH_SCAN` / `CONNECT`;
 * below that, scanning counts as a location request. iOS handles this through
 * the Info.plist usage string and needs nothing here.
 */
export async function ensureBlePermissions(): Promise<boolean> {
  if (Platform.OS !== 'android') return true;

  const apiLevel = typeof Platform.Version === 'number' ? Platform.Version : 0;

  if (apiLevel >= 31) {
    const statuses = await PermissionsAndroid.requestMultiple([
      PermissionsAndroid.PERMISSIONS.BLUETOOTH_SCAN,
      PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT,
    ]);
    const granted = PermissionsAndroid.RESULTS.GRANTED;
    return (
      statuses[PermissionsAndroid.PERMISSIONS.BLUETOOTH_SCAN] === granted &&
      statuses[PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT] === granted
    );
  }

  // Android 11 and below: scanning is gated on location.
  const location = await PermissionsAndroid.request(
    PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
  );
  return location === PermissionsAndroid.RESULTS.GRANTED;
}
